src/platform/imx8/include/platform/drivers/interrupt.h

```diff
@@ -60,7 +60,9 @@
 
 /* 32 HW interrupts + 8 IRQ_STEER lines each with 64 interrupts */
 #define PLATFORM_IRQ_HW_NUM	XCHAL_NUM_INTERRUPTS
-#define PLATFORM_IRQ_CHILDREN	64
+#define PLATFORM_IRQ_CHILDREN	64 /* Each cascaded struct covers 64 IRQs */
+/* IMX: Covered steer IRQs are modulo-64 aligned. [0-63] are useless. */
```

Comment:
I'm not sure about 32-63... Can you explain?

Reply:
Yes sorry I had missed your question. You are correct when asking for clarification.

We can align the first child on value 0, or like I did in my last commit, on 64 (because [0-63] are not relevant on HIFI4). Or we could even align on an higher multiple of 64 and spare lot of RAM by reducing the number of allocated cascaded structs (on IMX8 QM and QXP at least).
 
I just have double verified this afternoon that SAI plays correctly with both configurations (0 and 64). Note that for 64, we need to remove the corresponding unused cascaded struct from the dsp_irq[] array, like currently done in the commit.

Actually I am not sure of what is the best to push on master branch. 0 is more simple and generic for all platforms of the future, but aligning on an higher multiple of 64 saves a lot of RAM. On  IMX8 QM and QXP we could align on 5*64, reducing the memory print by 5x824 bytes.

What do you think ?

Reply:
> Also with this edit IRQSTR 442 becomes SOF 410.
I am not sure of your point. So I have added the following trace in irq_register_child():
	trace_error(TRACE_CLASS_IRQ,
		    "irq_register_child: irq %d, hw_irq %d, irq_base %d",
		    irq, hw_irq, cascade->irq_base);
I got, for SAI:

dai_comp_trigger(), START
irq_register_child: irq 349, hw_irq 29, irq_base 320

Since it works fine for SAI I didn't check deeper. I'll be back on Tuesday then we can clarify this point.

Regards  :-)

Reply:
Hi Paul,
thinking it makes things more simple I just have pushed a new version of the commit with first child starting at 0 so that it's reduced to its simplest form.

The optimization to reduce the memory footprint by having child at other 64-multiples can be done anytime later.

About your point that IRQSTR 442 becomes SOF 410, I cannot see it. The SAI plays well on my side. 
For example the following traces from platform_init() :

	// for ESAI
	irqstr_int = 442;
	line = irqstr_int / IRQSTR_IRQS_PER_LINE;
	irq = irqstr_int % IRQSTR_IRQS_PER_LINE;
	
	sofirq = irqstr_get_sof_int(irqstr_int);

	trace_irq_error("%s irqstr_int %d, line %d, irq %d, sofirq %d, %s",
			__func__, irqstr_int, line, irq, sofirq, irq_name_irqsteer[line] );

do generate that output:

	platform_interrupt_init` irqstr_int 442, line 6, irq 58, sofirq 442, irqsteer6

Don't hesitate to point me to the branch where the problem happens so I can have a look. 

Regards

Reply:
If you use `irqstr_get_sof_int` those shifts will not break any code. It will still play fine even if the shift occurs because this function is made robust against those shifts (it queries the SOF data structures itself to get the SOF interrupt number).

Well if you packed and removed the first 64 interrupts (outright remove them from the array) the next would come where the removed ones have come, so I think there would be a shift of 64 in the SOF IRQ numbers even. But the thing is that shift is accounted for, the thing is does it actually matter?

Having FIRST_CHILD be set to 0 means that there is an overlap between IRQ_STR and the Xtos interrupts, and the latter will shadow the first 32 IRQ_STR interrupts (which I believe is in fact the intended behavior in this case).

Just be aware that even with this change you shouldn't remove that function and just use the interrupt numbers directly -- future changes to interrupt handling both in our code and in generic SOF code might cause trouble in this sense.